### ISSUE TITLE
`check_type` fails with `TypedDict | None` and/or `Optional[TypedDict]`

### ISSUE THREAD

**Issue by ksquarekumar** [agronholm/typeguard#346]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### Typeguard version

3.0.2

### Python version

3.10.9

### What happened?

Expected `check_type` to pass, but it fails with the error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[13], line 1
----> 1 check_type(None, MyTypedDict | None)

File [~/Documents/git/venv/lib/python3.10/site-packages/typeguard/_functions.py:74](https://file+.vscode-resource.vscode-cdn.net/Users/user/Documents/git/~/Documents/git/.venv/lib/python3.10/site-packages/typeguard/_functions.py:74), in check_type(value, expected_type)
     72 memo = TypeCheckMemo(frame.f_globals, frame.f_locals)
     73 try:
---> 74     check_type_internal(value, expected_type, memo)
     75 except TypeCheckError as exc:
     76     exc.append_path_element(qualified_name(value, add_class_prefix=True))

File [~/Documents/git/.venv/lib/python3.10/site-packages/typeguard/_checkers.py:676](https://file+.vscode-resource.vscode-cdn.net/Users/user/Documents/git/~/Documents/git/.venv/lib/python3.10/site-packages/typeguard/_checkers.py:676), in check_type_internal(value, annotation, memo)
    674     checker = lookup_func(origin_type, args, extras)
    675     if checker:
--> 676         checker(value, origin_type, args, memo)
    677         return
    679 if not isinstance(value, origin_type):

File [~/Documents/git/.venv/lib/python3.10/site-packages/typeguard/_checkers.py:394](https://file+.vscode-resource.vscode-cdn.net/Users/user/Documents/git/~/Documents/git/.venv/lib/python3.10/site-packages/typeguard/_checkers.py:394), in check_uniontype(value, origin_type, args, memo)
    392 for type_ in args:
    393     try:
--> 394         check_type_internal(value, type_, memo)
    395         return
    396     except TypeCheckError as exc:
...
--> 242 existing_keys = frozenset(value)
    243 extra_keys = existing_keys - declared_keys
    244 if extra_keys:

TypeError: 'NoneType' object is not iterable
```

### How can we reproduce the bug?

```python
from typing import TypedDict
from typeguard import check_type

class _MyTypedDict(TypedDict, total=False):
    b: str

class MyTypedDict(_MyTypedDict, TypedDict):
    a: str

check_type(None, MyTypedDict | None)
```

or 

```python
from typing import TypedDict, Optional
from typeguard import check_type

class _MyTypedDict(TypedDict, total=False):
    b: str

class MyTypedDict(_MyTypedDict, TypedDict):
    a: str

check_type(None, Optional[MyTypedDict])
```

**Comment by agronholm** [comment]
I can't reproduce this on `master`. Could you verify that it works with the latest master?

**Comment by ksquarekumar** [comment]
yeah it works with `master` 👍🏽

**Comment by ksquarekumar** [comment]
Additionally, I have code that uses typer for the `cli` parts and on `master` and all of them fail on `assert result.exit_code == 0` when testing with `pytest` and `typer.testing.CliRunner`. So this makes `master` unusable for me. 

and I have since tried to use `suppress_type_checks` for the cases where `3.0.2` was failing, but this also has an issue, i.e.
I have also noticed that `suppress_type_checks()` fails to close the context in `3.0.2` so tests running after the suppressed test fail with weird errors on mocks getting all Nones.

**Comment by agronholm** [comment]
I'm a bit confused. Is there another bug you would like to report?

**Comment by agronholm** [comment]
> I have also noticed that suppress_type_checks() fails to close the context when attaching inside a staticmethod function, so tests running after the suppressed test fail with weird errors on mocks returning all Nones.

This was fixed in v4.0.0rc1.
